### PR TITLE
Add state comparison to router

### LINKED ISSUE #38267: SSG pages re-render if middleware is being used
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

   Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Home
    Binaries:
      Node: 14.18.0
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 12.2.1-canary.2
      eslint-config-next: 12.2.0
      react: 18.2.0
      react-dom: 18.2.0

### What browser are you using? (if relevant)

Chrome 103.0.5060.66

### How are you deploying your application? (if relevant)

next start, Vercel

### Describe the Bug

By using the new stable middleware.js at the root of the project, all pages will render twice. With the old _middleware.js, there was no similar problem. 

### Expected Behavior

Pages render only once

### Link to reproduction

https://github.com/imangm/nextjs-middleware-rerender

### To Reproduce

1. Clone the repo
2. Run `npm run build`
3. Run `npm run start`
4. In Chrome DevTools, go to Network tab -> Network conditions
5. Turn on **Disable cache**, also set Network throttling to **Slow 3G** to see the issue clearly
6. Open the project in chrome
7. Once the page loads, quickly scroll down, once the page is loaded once, it will load again and will automatically scroll to top
8. In the console, you can see **render _app** is logged twice
9. Also if you run the project using `npm run dev`, you can check the chrome console to see the re-render
10. A live version of the repo is available here: https://next-middleware-rerender.vercel.app/

### PR DESCRIPTION
**Pull request by shuding**
Lands #37431 again, but it only solves the re-render issue completely for the middleware case (closes #38267), not the `rewrites` case (#37139).

For `rewrites`, the blocker is `isReady` always being `false` initially and to match the markup on hydration we can't simply work around it without a re-render. Need to come up with another fix.

## Bug

- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `pnpm lint`
- [ ] The examples guidelines are followed from [our contributing doc](https://github.com/vercel/next.js/blob/canary/contributing.md#adding-examples)

### CODE REVIEW

**Review by SuttonJack — DISMISSED** [review]

**Comment by shuding** [comment]
Hey @11koukou I'm gonna fix the tests and get this merged as soon as possible.

**Review by ijjk — COMMENTED** [review]
It seems like we may want to address this by ensuring `isReady` is correct when initializing the router, possibly we can break out the resolving logic for rewrites and such so that we can call this before creating the router in `client/index.tsx` 🤔

File: packages/next/shared/lib/router/router.ts [hunk 1]
```diff
@@ -1474,46 +1476,63 @@ export default class Router implements BaseRouter {
...
+
+      // for query updates we can skip it if the state is unchanged and we don't
+      // need to scroll
+      // https://github.com/vercel/next.js/issues/37139
+      const canSkipUpdating =
+        (options as any)._h &&
+        !upcomingScrollState &&
+        compareRouterStates(upcomingRouterState, this.state) &&
+        !readyStateChange &&
+        !localeChange
```
Comment (ijjk) [inline packages/next/shared/lib/router/router.ts:1502]:
```suggestion
        !readyStateChange &&
        !localeChange &&
        compareRouterStates(upcomingRouterState, this.state) 
```
Should we keep the state comparison last so we only do the comparison when necessary?

**Comment by 11koukou** [comment]
> Hey @11koukou I'm gonna fix the tests and get this merged as soon as possible.

Thank you very much

File: packages/next/shared/lib/router/router.ts (hunk 1 above)
Comment (shuding) [reply packages/next/shared/lib/router/router.ts:1502]: Makes sense!

**Review by ijjk — APPROVED** [review]

**Pull request merged**